### PR TITLE
Return (response, error) #2

### PR DESCRIPTION
**Pull request by tamird**
More work toward #1400.

### CODE REVIEW

File: storage/range.go [hunk 1]
```diff
@@ -507,19 +511,35 @@ func (r *Range) SetLastVerificationTimestamp(timestamp proto.Timestamp) error {
...
 	} else if proto.IsReadOnly(args) {
-		defer trace.Epoch("read path")()
-		return r.addReadOnlyCmd(ctx, args, reply)
+		defer trace.Epoch("read-only path")()
+		reply, err = r.addReadOnlyCmd(ctx, args)
+	} else if proto.IsWrite(args) {
+		defer trace.Epoch("read-write path")()
+		reply, err = r.addWriteCmd(ctx, args, nil)
+	} else {
+		panic("incomplete match")
```
Comment (bdarnell) [inline storage/range.go]: This is a strange error message. How about `panic(fmt.Sprintf("don't know how to handle command %T", args))`

File: storage/range.go [hunk 2]
```diff
@@ -567,66 +587,62 @@ func (r *Range) endCmd(cmdKey interface{}, args proto.Request, err error, readOn
...
 	}
 
-	switch args.(type) {
+	var err error
+	var reply proto.Response
+	switch tArgs := args.(type) {
 	case *proto.AdminSplitRequest:
-		r.AdminSplit(args.(*proto.AdminSplitRequest), reply.(*proto.AdminSplitResponse))
+		var resp proto.AdminSplitResponse
+		resp, err = r.AdminSplit(tArgs)
```
Comment (bdarnell) [inline storage/range.go]: There's no reason that `err` has to be set in the outer scope. I would make this 

    resp, err := r.AdminSplit(tArgs)
    return &resp, err

I have some doubts about returning the protos as value types from the individual methods when they're pointers everywhere else. In this case we could do `return r.AdminSplit(tArgs)` if the return types were the same (although maybe that invites trouble as a nil `*proto.AdminSplitResponse` could turn into a non-nil `proto.Response`).

File: storage/range.go [hunk 3]
```diff
@@ -507,19 +511,35 @@ func (r *Range) SetLastVerificationTimestamp(timestamp proto.Timestamp) error {
...
+	} else {
+		panic("incomplete match")
 	}
-	return r.addWriteCmd(ctx, args, reply, nil)
+
+	if reply != nil {
+		gogoproto.Merge(call.Reply, reply)
+	}
+
+	if err != nil {
```
Comment (bdarnell) [inline storage/range.go]: The assertion that `Reply.Header().GoError()` is nil should move along with the call to `SetGoError`

**Comment by bdarnell** [comment]
LGTM

File: storage/range.go (hunk 1 above)
Comment (tamird) [inline storage/range.go]: done

File: storage/range.go (hunk 2 above)
Comment (tamird) [inline storage/range.go]: > although maybe that invites trouble as a nil `*proto.AdminSplitResponse` could turn into a non-nil `proto.Response`

This was my worry (and I got bitten by unintentionally returning null pointers during this refactor). A small price for a small increase in type safety.

I've implemented your suggestion.

File: storage/range.go (hunk 3 above)
Comment (tamird) [inline storage/range.go]: done

File: storage/range.go [hunk 4]
```diff
@@ -914,31 +912,28 @@ func (r *Range) applyRaftCommandInBatch(ctx context.Context, index uint64, origi
...
-			return batch, newReplicaCorruptionError(util.Errorf("could not read from response cache"), err)
-		} else if cachedReply != nil {
-			// TODO(tamird): this shouldn't be needed
-			gogoproto.Merge(reply, cachedReply)
+			return batch, reply, newReplicaCorruptionError(util.Errorf("could not read from response cache"), err)
+		} else if reply != nil {
 			if log.V(1) {
 				log.Infoc(ctx, "found response cache entry for %+v", args.Header().CmdID)
 			}
+			// TODO(tamird): move this into the response cache itself
```
Comment (tschottdorf) [inline storage/range.go]: `itself.`

File: storage/range_command.go [hunk 5]
```diff
@@ -849,34 +849,30 @@ func (r *Range) AdminSplit(args *proto.AdminSplitRequest, reply *proto.AdminSpli
 	if len(splitKey) == 0 {
 		snap := r.rm.NewSnapshot()
 		defer snap.Close()
-		var err error
-		if splitKey, err = engine.MVCCFindSplitKey(snap, desc.RaftID, desc.StartKey, desc.EndKey); err != nil {
-			reply.SetGoError(util.Errorf("unable to determine split key: %s", err))
-			return
+		foundSplitKey, err := engine.MVCCFindSplitKey(snap, desc.RaftID, desc.StartKey, desc.EndKey)
```
Comment (tschottdorf) [inline storage/range_command.go]: `foundSplitKey` sounds like it would be a `bool`.

Comment (tamird) [inline storage/range_command.go]: Yeah, not the best name. Do you have a suggestion?

Comment (tschottdorf) [inline storage/range_command.go]: not really and doesn't matter. Best I can think of is

```go
splitKey := func() proto.Key {
    if len(args.SplitKey) > 0 {
        return args.SplitKey
    }
    // rest
}
```

**Pull request merged**